docs/_blog/bdk_core_pt1.md

```diff
@@ -0,0 +1,322 @@
+
+So what is the most general syncing mechanism that solves these problems? These are the things I
+think it has to do regardless of where the blockchain data comes from or how it's stored:
+
+1. Generate and store addresses.
+2. Index transaction data. e.g. transaction outputs we own, when/if they were spent etc.
+3. Keep track of which addresses have been given out and which have been used.
+4. Be able to "roll back" our view of the above data if a re-org makes some of it stale.
+5. Keeping track of transactions related our addresses in our mempool.
+
+Let's talk about how to implement a mechanism that does all that.
+
+### How to store and index transactions
+
+Different persistent storage backends have different APIs and their own indexing strategies. That's
+why the [`Database`] trait exists in BDK to make a clean API to the different storage engines. It's
+important to note that the database in BDK only holds public data that could always be retrieved
+from the chain. It's just a cache. Despite this we support different backends. Right now it is a a
+lot of work to add a new index to the data since you have add it to every backend and you might have
+to apply schema changes (we still [don't have a standard approach to
+this](https://github.com/bitcoindevkit/bdk/issues/359)).
+
+Thomas Eizinger [suggested](https://github.com/bitcoindevkit/bdk/issues/165#issuecomment-1047483895)
+doing everything in memory and only writing to persistent storage when it was convenient. It took me
+some time but I came around to this idea. It would allow us to get rid of the `Database` trait (at
+least at the `bdk_core` level) and greatly simplify what the persistent storage layer has to do.
+Whenever the data is loaded from persistent storage we can just do the indexing in memory and
+present it to the application.
+
+*But wait! Wouldn't this mean we'd use way more memory than we need to?* Yes but memory is cheap.
+Consider that if we say the average transaction size is 300 bytes then with all our indexes each
+transaction might cost 1kb of memory (pessimistically). This means we could index one thousand
+transactions in a single megabyte! My iPhone has 4gb of memory so it could index a million
+transactions with plenty of memory to spare. *But what if some users can't afford an iPhone?* Then
+they also couldn't have afforded to have made a million Bitcoin transactions! *But what about memory
+constrained devices like hardware wallets!?* Those devices typically don't store and retrieve
+transactions. They're usually just signing devices. Perhaps one day someone will build a memory
+constrained device that needs to do this work but until then I think this is a fine approach to
+take.
+
+For now I'm calling this thing that does the in memory indexing of transactions related to a single
+descriptor a `DescriptorTracker`. Here's a diagram that communicates how I imagine it relates to the
+other components.
+
+![](./bdk_core_pt1/descriptor-tracker.jpg)
```

Comment:
Where is the diagram source code?

Reply:
it comes from miro.com so I don't have source.